### PR TITLE
Unit test task

### PR DESCRIPTION
**Pull request by nehayward**
**What does this pull request do?**
Initial pass at unit test for aws provider.

**How should this be tested?**
Include steps to test intended functionality.

**Checklist**
- [x] Unit tests

closes #
(reference the issue(s) this pull request closes)

### CODE REVIEW

File: api/provider/aws/test_aws/task_create_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	"github.com/golang/mock/gomock"
+	provider "github.com/quintilesims/layer0/api/provider/aws"
+	"github.com/quintilesims/layer0/api/tag"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config/mock_config"
+	"github.com/quintilesims/layer0/common/models"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTaskCreate(t *testing.T) {
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: This file/function suggest a `Create` test, but this is a `List` test

File: api/provider/aws/test_aws/task_create_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// ensure we filter out clusters that don't belong to our instance
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: we don't need to assert this behavior in this test. This is handled by environment list test

File: api/provider/aws/test_aws/task_create_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+	mockAWS.ECS.EXPECT().
+		ListClustersPages(gomock.Any(), gomock.Any()).
+		Do(listClusterPagesFN).
+		Return(nil)
+
+	// ensure we filter out tasks that don't belong to our instance
+	listTaskPagesFN := func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
+		taskARNs := []*string{
+			aws.String("arn:aws:ecs:region:012345678910:task/l0-test-tsk_id1"),
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: these are not what task arns look like, and they also don't match the arns you specified in tags above. They need to match because every arn returned here that doesn't have a matching tag gets filtered out. I'm not sure how this test is passing tbh because it should return an empty slice of task summaries with the current way it is setup.

File: api/provider/aws/test_aws/task_create_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// ensure we filter out clusters that don't belong to our instance
+	// TODo: Make Helper?
+	listClusterPagesFN := func(input *ecs.ListClustersInput, fn func(output *ecs.ListClustersOutput, lastPage bool) bool) error {
+		clusterARNs := []*string{
+			aws.String("arn:aws:ecs:region:012345678910:cluster/l0-test-env_id1"),
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: you need a way to relate the clusters returned here to the `ListTaskPages` below. In the actual implementation, we call `ListTaskPages` for **each** cluster (not just once)

File: api/provider/aws/test_aws/task_delete_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		{
+			EntityID:   "tsk_id",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn1",
+		},
+		{
+			EntityID:   "tsk_id",
+			EntityType: "task",
+			Key:        "id",
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: `environment_id`

File: api/provider/aws/test_aws/task_create_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		{
+			EntityID:   "tsk_id2",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn2",
+		},
+		{
+			EntityID:   "tsk_id2",
+			EntityType: "task",
+			Key:        "id",
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: `environment_id`

File: api/provider/aws/test_aws/task_create_test.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn1",
+		},
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "id",
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: `environment_id`

File: api/provider/aws/test_aws/task_delete_test.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	stopTaskInput := &ecs.StopTaskInput{}
+	stopTaskInput.SetCluster("l0-test-env_id")
+	stopTaskInput.SetTask("l0-test-tsk_id")
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: We pass in the task arn, not the task id into `StopTask` (https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/task_delete.go#L26). I'm not sure how this test is passing

File: api/provider/aws/test_aws/task_delete_test.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+	mockAWS.ECS.EXPECT().
+		StopTask(stopTaskInput).
+		Return(nil, nil)
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: `Return(&ecs.StopTaskOutput{}, nil)`

File: api/provider/aws/test_aws/task_delete_test.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	err := target.Delete("tsk_id")
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: `if err := ...; err != nil { ... }`

File: api/provider/aws/test_aws/task_delete_test.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	// todo: setup helper for config
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	mockAWS.ECS.EXPECT().
+		StopTask(gomock.Any()).
+		Return(nil, awserr.New("TaskDoesNotExist", "", nil))
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: This doesn't check that error code, it checks the message: https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/task_delete.go#L47

File: api/provider/aws/test_aws/task_list_test.go [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		Do(listClusterPagesFN).
+		Return(nil)
+
+	for _, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
+		listTaskPagesFN := func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
+			taskARNs := []*string{
+				aws.String("arn:aws:ecs:region:012345678910:task/arn1"),
+				aws.String("arn:aws:ecs:region:012345678910:task/arn2"),
+			}
+			output := &ecs.ListTasksOutput{}
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: [nit] newline after closing bracket

File: api/provider/aws/test_aws/task_list_test.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+		return nil
+	}
+
+	mockAWS.ECS.EXPECT().
+		ListClustersPages(gomock.Any(), gomock.Any()).
+		Do(listClusterPagesFN).
+		Return(nil)
+
+	for _, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
+		listTaskPagesFN := func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: This is on the right track, but a little wonky if you think about it. The scenario setup here is we have 2 environments in our layer0 instance, and each environment has 2 tasks but they have the exact same arns. This could never happen in real life (each task's arn is unique), so we should consider changing this a bit, e.g. cluster1 returns arn1, cluster2 returns arn2

File: api/provider/aws/test_aws/task_list_test.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+				aws.String("arn:aws:ecs:region:012345678910:task/arn2"),
+			}
+			output := &ecs.ListTasksOutput{}
+			output.SetTaskArns(taskARNs)
+			fn(output, true)
+
+			return nil
+		}
+
+		for _, status := range []string{ecs.DesiredStatusRunning, ecs.DesiredStatusStopped} {
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: 👍 nice

File: api/provider/aws/test_aws/task_read_test.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn1",
+		},
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "id",
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: `environment_id`

File: api/provider/aws/test_aws/task_read_test.go [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// environments' asg names are the same as the fq environment id
+	describeASGInput := &autoscaling.DescribeAutoScalingGroupsInput{}
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: this doesn't make any sense to me - we don't describe autoscaling groups in a task read?

File: api/provider/aws/test_aws/task_read_test.go [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+	lc := &autoscaling.LaunchConfiguration{}
+	lc.SetLaunchConfigurationName("lc_name")
+	lc.SetInstanceType("m3.small")
+	lc.SetImageId("ami_id")
+
+	describeLCOutput := &autoscaling.DescribeLaunchConfigurationsOutput{}
+	describeLCOutput.SetLaunchConfigurations([]*autoscaling.LaunchConfiguration{lc})
+
+	taskARNs := []*string{
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: we should only be expecting the task arn from the tag above

File: api/provider/aws/test_aws/task_read_test.go [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+		aws.String("arn:aws:ecs:region:012345678910:task/l0-bad-tsk_id1"),
+		aws.String("arn:aws:ecs:region:012345678910:task/bad2"),
+	}
+
+	input := &ecs.DescribeTasksInput{}
+	// input.SetCluster()
+	input.SetTasks(taskARNs)
+
+	output := &ecs.DescribeTasksOutput{}
+	// output.SetTasks()
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: This definitely needs the output set https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/task_read.go#L66

File: api/provider/aws/test_aws/task_read_test.go [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+		Return(output, nil)
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	result, err := target.Read("tsk_id")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// todo: Links
+	expected := &models.Task{}
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: You need a fully populated model (including a non-empty `Containers` field)

File: api/provider/aws/test_aws/task_delete_test.go [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+		StopTask(stopTaskInput).
+		Return(nil, nil)
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	err := target.Delete("tsk_id")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Len(t, tagStore.Tags(), 0)
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: shouldn't the environment's name tag still exist?

File: api/provider/aws/test_aws/task_list_test.go [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	//[l0-nick-demo91e6e283 l0-nick-api l0-nick-demoenve36e9 l0-nick-endtoenb8169]
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: remove debug stuff

File: api/provider/aws/test_aws/task_list_test.go [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+		output := &ecs.ListClustersOutput{}
+		output.SetClusterArns(clusterARNs)
+
+		fn(output, true)
+
+		return nil
+	}
+
+	mockAWS.ECS.EXPECT().
+		ListClustersPages(gomock.Any(), gomock.Any()).
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: `ListClustersPages(&ecs.ListClustersPagesInput{}, gomock.Any()).`

File: api/provider/aws/test_aws/task_read_test.go [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+
+	taskARNs := []*string{
+		aws.String("arn:aws:ecs:region:012345678910:task/l0-test-tsk_id1"),
+		aws.String("arn:aws:ecs:region:012345678910:task/l0-test-tsk_id2"),
+		aws.String("arn:aws:ecs:region:012345678910:task/l0-bad-tsk_id1"),
+		aws.String("arn:aws:ecs:region:012345678910:task/bad2"),
+	}
+
+	input := &ecs.DescribeTasksInput{}
+	// input.SetCluster()
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: needs to be set

File: api/provider/aws/test_aws/task_delete_test.go (hunk 20 above)
Comment (jparsons04) [reply api/provider/aws/test_aws/task_delete_test.go]: I see only Task entity type tags defined above.

File: api/provider/aws/test_aws/task_delete_test.go [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+	stopTaskInput.SetTask("arn:aws:ecs:region:012345678910:task/arn1")
+
+	mockAWS.ECS.EXPECT().
+		StopTask(stopTaskInput).
+		Return(nil, nil)
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	err := target.Delete("tsk_id")
+	if err != nil {
+		// t.Fatal(err)
```
Comment (jparsons04) [inline api/provider/aws/test_aws/task_delete_test.go]: This also needs the `if err := ...; err != nil { ... }` pattern. Also, it seems that this should be uncommented.

Comment (nehayward) [reply api/provider/aws/test_aws/task_delete_test.go]: This is something I wanted to ask about if you look at the task delete it reads the aws error message. I’m wondering if I should do the same or if we should change both. 
@zpatrick

Comment (jparsons04) [reply api/provider/aws/test_aws/task_delete_test.go]: Ah, I see now. So, yes, since this test is testing for idempotence, we're mainly trying to test the graceful handling of an error returned from the SDK. For instance, the SDK helper function [deleteLoadBalancer](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/load_balancer_delete.go#L54) checks for the presence of a "NoSuchEntity" error code from the SDK and exits gracefully. Likewise [my EXPECT() call](https://github.com/quintilesims/layer0/pull/410/files#diff-ed65c448f68e560743c6c2a4999f9b22R93) ensures that this is happened.

The other helper functions that are called have similar error code checks. In your case, the idempotent check [looks for a specific string in the error message](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/task_delete.go#L47). @zpatrick has a good example of how to handle this [here](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/test_aws/environment_delete_test.go#L101)

File: api/provider/aws/test_aws/task_create_test.go [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	req := models.CreateTaskRequest{
+		DeployID:           "dpl_id",
+		EnvironmentID:      "env_id",
+		TaskName:           "tsk_name",
+		ContainerOverrides: []models.ContainerOverride{},
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_create_test.go]: this should be a non zero-value/empty object so we make sure we map these params correctly

File: api/provider/aws/test_aws/task_delete_test.go [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+		}
+	}
+
+	stopTaskInput := &ecs.StopTaskInput{}
+	stopTaskInput.SetCluster("l0-test-env_id")
+	stopTaskInput.SetTask("arn:aws:ecs:region:012345678910:task/arn")
+
+	mockAWS.ECS.EXPECT().
+		StopTask(stopTaskInput).
+		Return(nil, nil)
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: `Return(&ecs.StopTaskOutput{}, nil)`

File: api/provider/aws/test_aws/task_list_test.go [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+
+		taskARNs := []*string{
+			aws.String("arn:aws:ecs:region:012345678910:task/arn2"),
+			aws.String("arn:aws:ecs:region:012345678910:task/arn1"),
+		}
+
+		taskARNs = append(taskARNs[:i], taskARNs[i+1:]...)
+
+		listTaskPagesFN := func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
+
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: [nit] remove newline

File: api/provider/aws/test_aws/task_list_test.go [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+		Return(nil)
+
+	for i, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
+
+		taskARNs := []*string{
+			aws.String("arn:aws:ecs:region:012345678910:task/arn2"),
+			aws.String("arn:aws:ecs:region:012345678910:task/arn1"),
+		}
+
+		taskARNs = append(taskARNs[:i], taskARNs[i+1:]...)
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: this was hard to parse, maybe we can simplify a bit:
```
for i, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
    taskARNs := []string{"arn:aws:ecs:region:012345678910:task/arn1"}
    if i == 1 {
        taskARNs = []string{"arn:aws:ecs:region:012345678910:task/arn2"}
    }
 
    ...
}
```

File: api/provider/aws/test_aws/task_read_test.go [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+
+	expected := &models.Task{
+		TaskID:        "tsk_id",
+		TaskName:      "tsk_name",
+		EnvironmentID: "env_id",
+		DeployID:      "dpl_id",
+		DeployName:    "dpl_name",
+		DeployVersion: "version",
+		Status:        "RUNNING",
+		Containers:    []models.Container{},
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: we need a zero-value/nil field here to ensure we map correctly.

File: api/provider/aws/test_aws/task_list_test.go (hunk 28 above)
Comment (nehayward) [reply api/provider/aws/test_aws/task_list_test.go]: Yeah I might have over complicated it.

File: api/provider/aws/test_aws/task_delete_test.go [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	mockAWS.ECS.EXPECT().
+		StopTask(gomock.Any()).
+		Return(nil, awserr.New("TaskDoesNotExist", "Taks 'tsk_id' does not exist", nil))
```
Comment (diemonster) [inline api/provider/aws/test_aws/task_delete_test.go]: [nit] typo "Taks"

File: api/provider/aws/test_aws/task_list_test.go [hunk 31]
```diff
@@ -0,0 +1,173 @@
...
+
+		return nil
+	}
+
+	mockAWS.ECS.EXPECT().
+		ListClustersPages(&ecs.ListClustersInput{}, gomock.Any()).
+		Do(listClusterPagesFN).
+		Return(nil)
+
+	var taskARNs = map[string]map[string][]*string{}
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: I think using a 4-dimensional map is overthinking this. We expect 4 calls and want to return a different arn each time - sounds like this can just be accomplished with a `switch` statement within whatever function you pass to `.Do(...)` (see below)

File: api/provider/aws/test_aws/task_list_test.go [hunk 32]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	for _, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
+		for _, status := range []string{ecs.DesiredStatusRunning, ecs.DesiredStatusStopped} {
+			listTasksInput := &ecs.ListTasksInput{}
+			listTasksInput.SetCluster(environmentID)
+			listTasksInput.SetDesiredStatus(status)
+			listTasksInput.SetStartedBy("test")
+
+			mockAWS.ECS.EXPECT().
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]:
```go
newListTaskPagesFN := func(i int) func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
    return func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
        var arn string
        switch i {
        case 0:
            arn = "../arn1"
        ...
        default:
            arn = ".../arn4"
        }

        output := &ecs.ListTasksOutput{}
        output.SetTaskArns([]*string{arn})
        fn(output, true)
 
        return nil
    })
}

for i := range  environments {
    for j := range statuses {
        mockAWS.ECS.EXPECT().
            ListTasksPages(listTasksInput, gomock.Any()).
            Do(newListTaskPagesFN(i+j)).
            Return(nil)
    }
}
```

File: api/provider/aws/test_aws/task_read_test.go [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+			EntityID:   "dpl_id",
+			EntityType: "deploy",
+			Key:        "name",
+			Value:      "dpl_name",
+		},
+		{
+			EntityID:   "dpl_id",
+			EntityType: "deploy",
+			Key:        "version",
+			Value:      "version",
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: this should be a different value than the key

File: api/provider/aws/test_aws/task_read_test.go [hunk 34]
```diff
@@ -0,0 +1,108 @@
...
+		Return(describeTaskOutput, nil)
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	result, err := target.Read("tsk_id")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	container := models.Container{
+		ContainerName: "container",
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: We need to check the `Status` and `ExitCode` fields

File: api/provider/aws/test_aws/task_list_test.go (hunk 32 above)
Comment (zpatrick) [reply api/provider/aws/test_aws/task_list_test.go]: you could avoid the `switch` statement since each arn is just a prefix + integer suffix, e.g. 
```
arn := fmt.Sprintf("aws:arn:..../arn%d", i+1)
```

File: api/provider/aws/common.go [hunk 35]
```diff
@@ -82,7 +82,7 @@ func lookupTaskDefinitionARNFromDeployID(store tag.Store, deployID string) (stri
 func lookupEntityEnvironmentID(store tag.Store, entityType, entityID string) (string, error) {
 	tags, err := store.SelectByTypeAndID(entityType, entityID)
 	if err != nil {
-		return "", err
+		return "ts", err
```
Comment (jparsons04) [inline api/provider/aws/common.go]: This seems curious to me. I didn't see any references to the string `"ts"` for each reference I could find for this function. Is this string being explicitly tested somewhere that I'm not seeing?

File: api/provider/aws/test_aws/task_list_test.go [hunk 36]
```diff
@@ -0,0 +1,154 @@
...
+			Key:        "name",
+			Value:      "env_name1",
+		},
+		{
+			EntityID:   "env_id2",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name2",
+		},
+	}
```
Comment (jparsons04) [inline api/provider/aws/test_aws/task_list_test.go]: I'm curious what happens in the case where a layer0 environment not named `test` (e.g. `l0-someotherenvironment-env_id1`) is included in the tags db). Would list still return only the clusters with the environment name `test`?

File: api/provider/aws/common.go (hunk 35 above)
Comment (nehayward) [reply api/provider/aws/common.go]: Good catch. Hmm I don't recall adding this. I don't think it's necessary.

File: api/provider/aws/test_aws/task_list_test.go (hunk 36 above)
Comment (nehayward) [reply api/provider/aws/test_aws/task_list_test.go]: So if we add something like this. 
```go
	{
		EntityID:   "env_id3",
		EntityType: "environment",
		Key:        "name",
		Value:      "env_name3",
	},
```
It'll still only return the list of tasks in "test" environment (set environment). At least in testing, however this functionality is achieved by cli via the `l0 task list --all`

File: api/provider/aws/task_delete.go [hunk 37]
```diff
@@ -13,12 +14,12 @@ import (
 func (t *TaskProvider) Delete(taskID string) error {
 	environmentID, err := lookupEntityEnvironmentID(t.TagStore, "task", taskID)
 	if err != nil {
-		return err
+		log.Printf("[WARN] Environment not found\n")
```
Comment (zpatrick) [inline api/provider/aws/task_delete.go]: We need to do type assertions on the error and ensure that the error code is `TaskDoesNotExist`. Only in that scenario do we `return nil` early (also, you forgot to `return nil` after logging) .

File: api/provider/aws/task_delete.go [hunk 38]
```diff
@@ -13,12 +14,12 @@ import (
...
 	}
 
 	taskARN, err := t.lookupTaskARN(taskID)
 	if err != nil {
-		return err
+		log.Printf("[WARN] Task not found\n")
```
Comment (zpatrick) [inline api/provider/aws/task_delete.go]: ditto

File: api/provider/aws/test_aws/task_delete_test.go [hunk 39]
```diff
@@ -0,0 +1,135 @@
...
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	mockAWS.ECS.EXPECT().
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_delete_test.go]: we shouldn't get this far if the tags don't exist, correct?

File: api/provider/aws/test_aws/task_list_test.go (hunk 36 above)
Comment (zpatrick) [reply api/provider/aws/test_aws/task_list_test.go]: @jparsons04: I think you're conflating environment ids with the layer0 instance name, e.g. `env_id1` vs. `test`

File: api/provider/aws/test_aws/task_list_test.go [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+	for i, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
+		for j, status := range []string{ecs.DesiredStatusRunning, ecs.DesiredStatusStopped} {
+			listTasksInput := &ecs.ListTasksInput{}
+			listTasksInput.SetCluster(environmentID)
+			listTasksInput.SetDesiredStatus(status)
+			listTasksInput.SetStartedBy("test")
+
+			mockAWS.ECS.EXPECT().
+				ListTasksPages(listTasksInput, gomock.Any()).
+				Do(newListTaskPagesFN(i + j)).
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: doesn't this create a `aws:arn.../arn0` on the first iteration? Seems like we should be testing that 4 tasks are returned: 2 for each environment

File: api/provider/aws/test_aws/task_read_test.go [hunk 41]
```diff
@@ -0,0 +1,112 @@
...
+		Status:        "status",
+		ExitCode:      0,
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: We should avoid using 0-values in tests; this should probably be a `1`

File: api/provider/aws/test_aws/task_list_test.go (hunk 40 above)
Comment (nehayward) [reply api/provider/aws/test_aws/task_list_test.go]: Yeah I'll fix that. I agree, is that just add 4 task tags and assert 4 exist.

File: api/provider/aws/task_delete.go (hunk 37 above)
Comment (nehayward) [reply api/provider/aws/task_delete.go]: Okay and this behavior is already done here right https://github.com/quintilesims/layer0/blob/e6cf881ab637f2f5cd57848acc639fa262251b74/api/provider/aws/task_delete.go#L48

Comment (zpatrick) [reply api/provider/aws/task_delete.go]: similar behavior, but you should be checking for an `*errors.ServerError` and checking the `.StatusCode`, not the message

File: api/provider/aws/test_aws/task_list_test.go [hunk 42]
```diff
@@ -0,0 +1,205 @@
...
+			output.SetTaskArns(arn)
+			fn(output, true)
+
+			return nil
+		}
+
+		return listTaskPagesFN
+	}
+
+	taskARNCount := 1
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_list_test.go]: [nit] could just used `i + j + 1`

File: api/provider/aws/test_aws/task_list_test.go (hunk 36 above)
Comment (jparsons04) [reply api/provider/aws/test_aws/task_list_test.go]: Yep, sorry. I think my original comment can be discarded.

File: api/provider/aws/test_aws/task_list_test.go (hunk 42 above)
Comment (nehayward) [reply api/provider/aws/test_aws/task_list_test.go]: Not quite, since it won't hit every case, then if you try to add 1 it just complicates things and still doesn't hit every case unless you apply some conditional.
0 + 0 = 0
0 + 1 = 1
1 + 0 = 1
1 + 1 = 2

File: api/provider/aws/task_delete.go [hunk 43]
```diff
@@ -13,11 +15,19 @@ import (
...
+		log.Printf("[WARN] Environment not found\n")
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.TaskDoesNotExist {
+			return nil
+		}
 		return err
 	}
 
 	taskARN, err := t.lookupTaskARN(taskID)
 	if err != nil {
+		log.Printf("[WARN] Task not found\n")
```
Comment (zpatrick) [inline api/provider/aws/task_delete.go]: This needs to be inside the `if err, ok` block

File: api/provider/aws/test_aws/task_list_test.go (hunk 42 above)
Comment (zpatrick) [reply api/provider/aws/test_aws/task_list_test.go]: oops, my bad!

File: api/provider/aws/test_aws/task_read_test.go [hunk 44]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	describeTaskInput := &ecs.DescribeTasksInput{}
+	describeTaskInput.SetCluster("l0-test-env_id")
+	describeTaskInput.SetTasks([]*string{aws.String("arn:aws:ecs:region:012345678910:task/arn")})
+
+	containerECS := &ecs.Container{}
+	containerECS.SetName("container")
+	containerECS.SetLastStatus("status")
+	containerECS.SetExitCode(0)
```
Comment (zpatrick) [inline api/provider/aws/test_aws/task_read_test.go]: let's set this to 1 so we aren't using 0-values

File: api/provider/aws/task_delete.go [hunk 45]
```diff
@@ -13,11 +15,19 @@ import (
 func (t *TaskProvider) Delete(taskID string) error {
 	environmentID, err := lookupEntityEnvironmentID(t.TagStore, "task", taskID)
 	if err != nil {
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.TaskDoesNotExist {
+			log.Printf("[WARN] Environment not found\n")
```
Comment (diemonster) [inline api/provider/aws/task_delete.go]: `task` not `environment`
